### ISSUE TITLE
Automatic implementation for abstract methods in concrete classes

### ISSUE THREAD

**Issue by dgalichet** [scalameta/metals-feature-requests#25]
**Is your feature request related to a problem? Please describe.**

Not related to a problem

**Describe the solution you'd like**

This could take place automatically or with quick action.

```scala
  trait Foo {
    def foo(i: Int): String
  }
// before
  val foo = new Foo {}

// after
  val foo = new Foo {
    def foo(i: Int): String = ???
  }
```

**Describe alternatives you've considered**

This could be triggered using completion (same way that for the exhaustive pattern match), or using a quick action.

**Additional context**

**Search terms**:

**Comment by olafurpg** [comment]
Thanks for reporting! Have you tried using "override def" completions?

![2019-05-06 12 06 58](https://user-images.githubusercontent.com/1408093/57219158-7b6f2e80-6ff7-11e9-914c-aebfc9b5d1f8.gif)

It would be nice to expose this functionality as a code action, which are currently blocked by #18

**Comment by dgalichet** [comment]
yes actually I'm using it completion. But it could be nice to have all abstract methods to be appended when you define a concrete class. An alternative, it could be to have a quick action that append all these methods in case of error because of missing implementations.

**Comment by tanishiking** [comment]
I tried to implement this feature and found that there's an obstacle for realizing this feature.

## Background
In my plan, metals will provide the automatic implementation on the response of [Code Action Request](https://microsoft.github.io/language-server-protocol/specification#textDocument_codeAction) whose `CodeActionParams#context.kind` is `CodeActionKind.QuickFix`, and the metals server needs to route the `Code Action Request` to appropriate code action based on its `CodeActionParams`.

The best (in robustness) way is routing based on `Diagnostic#code`, which is a unique identifier of the diagnostics.

- Actually, each diagnostic in TSServer (TypeScript) has a unique identifier, and typescript language server-client tell is there available quick fixes or not, based on the diagnostic identifier.
  - FYI: list of diagnostics and codes: https://github.com/microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json
  - [typescript language server-client requests available quick fixes for the diagnostics, based on diagnostic code](https://github.com/microsoft/vscode/blob/f4524551e98945a3025f6038d3d32500182ff167/extensions/typescript-language-features/src/features/quickFix.ts#L161-L172)
  - (If there are no codeActions available for the diagnostic, [vscode will show the message "No quick fixes available"](https://github.com/microsoft/vscode/blob/9089a79ecf76a25fccc7c232d80cdd374a7cc66e/src/vs/editor/contrib/hover/modesContentHover.ts#L544-L555))

## Problem
Although [Diagnostic in BSP support a diagnostic code](https://github.com/scalacenter/bsp/blob/faf8c0293af388fdf90766b61d29d54a9ca67da2/docs/bsp.md#publish-diagnostics), bloop [returns `null` for the `code` field of Diagnostic](https://github.com/scalacenter/bloop/blob/d43d6a4ade1a494359dabceeaaac90380a15256b/frontend/src/main/scala/bloop/logging/BspServerLogger.scala#L108) because scalac (and even dotty) doesn't have unique identifier for error messages.
As a result, metals have no clue about what the errors are published by the build server, and about what quick fix should metals provide.

## Solution
### Assign unique identifiers to all errors in scalac / dotty
As discussed in https://contributors.scala-lang.org/t/towards-better-error-messages-in-scalac/1470/31, if the compiler errors have a unique identifier, metals can search the error and available codeActions robustly.
However, it may take quite a long time for implementing this.

## Workarounds
### Specify the available codeActions based on the error message
Diagnostic in LSP offers some other information other than `code`, like `range`, `source`, and `message`. Maybe we can decide what is the appropriate codeAction for the diagnostic, based on the error message.

For instance, if a class extends a trait and it lacks some required implementations, scala 2.12.9 publish the following error message,
```
object creation impossible, since method ${method} in trait ${trait} of type ... is not defined
```
In this case, we can tell the proper quick fix for this diagnostic is implementing the missing method automatically, because the error message starts with `"object creation impossible"` and ends with `"is not defined"`.
However, this way depends heavily on the compiler implementation and VERY vulnerable to the change in error messages.

**Comment by tgodzik** [comment]
The other solution could be checking if the error is in a class/object and then finding out if there is something unimplemented. It should be possible to go through parent and collect all abstract methods to compare them with the ones inside the newly created class.
